cupyx/scipy/sparse/index.py

```diff
@@ -0,0 +1,419 @@
+            elif col.ndim == 1:
+                return self._get_intXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif isinstance(row, slice):
+            if isinstance(col, INT_TYPES):
+                return self._get_sliceXint(row, col)
+            elif isinstance(col, slice):
+                if row == slice(None) and row == col:
+                    return self.copy()
+                return self._get_sliceXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_sliceXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif row.ndim == 1:
+            if isinstance(col, INT_TYPES):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_arrayXslice(row, col)
+        else:  # row.ndim == 2
+            if isinstance(col, INT_TYPES):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                raise IndexError('index results in >2 dimensions')
+            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
+                # special case for outer indexing
+                return self._get_columnXarray(row[:, 0], col.ravel())
+
+        # The only remaining case is inner (fancy) indexing
+        row, col = cupy.broadcast_arrays(row, col)
+        if row.shape != col.shape:
+            raise IndexError('number of row and column indices differ')
+        if row.size == 0:
+            return self.__class__(cupy.atleast_2d(row).shape, dtype=self.dtype)
+        return self._get_arrayXarray(row, col)
+
+    def _validate_indices(self, key):
+        M, N = self.shape
+        row, col = _unpack_index(key)
+
+        if isintlike(row):
+            row = int(row)
+            if row < -M or row >= M:
+                raise IndexError('row index (%d) out of range' % row)
+            if row < 0:
+                row += M
+        elif not isinstance(row, slice):
+            row = self._asindices(row, M)
+
+        if isintlike(col):
+            col = int(col)
+            if col < -N or col >= N:
+                raise IndexError('column index (%d) out of range' % col)
+            if col < 0:
+                col += N
+        elif not isinstance(col, slice):
+            col = self._asindices(col, N)
+
+        return row, col
+
+    def _asindices(self, idx, length):
+        """Convert `idx` to a valid index for an axis with a given length.
+        Subclasses that need special validation can override this method.
+        """
+        try:
+            x = cupy.asarray(idx, dtype="int32")
+        except (ValueError, TypeError, MemoryError):
+            raise IndexError('invalid index')
+
+        if x.ndim not in (1, 2):
+            raise IndexError('Index dimension must be <= 2')
+
+        if x.size == 0:
+            return x
+
+        if x.max().item() > length:
```

Comment:
This line requires device synchronization.
Generally, we cannot detect out-of-range-errors without looking at actual data. How about ignoring these kinds of errors as ndarray indexing-routines are doing?

Reply:
I had originally written a custom kernel to verify all these bounds with only a single synchronization. I removed it, though, to cut down on the number of custom kernels. If we don't throw the exceptions, I would just need to test whether any values are less than 0. Does that sound reasonable?